### DISCUSSION TITLE
Role grant + revoke conditions

### DISCUSSION THREAD

**Discussion by mark-ignacio** [member-gentei/member-gentei#335] [Ideas]
Based on some feedback, I plan to implement role grant and revoke conditions, which enables (1) support for "read the rules" roles and (2) arbitrary conditions based on the presence or absence of other roles. This new feature would be available through the `/gentei-if` command.

As currently planned, Gentei will evaluate these rules on every relevant role change.

**tl;dr**
* `/gentei-if set only:@roleyouneed not:@roleyoucannothave` to control role grants and revokes
* `/gentei-if test only:[...] not:[...] has:@big,@list,@of,@roles` to test rules w/o needing a test server or user
* `/gentei info` and `/gentei check` will contain the rules and how they apply to the user running it.

## Role grant/revoke conditions

`/gentei-if set` will be a new command + subcommand with 2 options:
* `only:` to *only* grant the membership role if a user holds one of these roles
* `not:` to *not* grant the membership role or remove a role if a user holds one of these roles

Gentei will react to role changes accordingly, skipping membership role grants evaluations 

Some use cases in mind:
* [Community Onboarding](https://support.discord.com/hc/en-us/articles/11074987197975-Community-Onboarding-FAQ) and [carl-bot reaction roles](https://docs.carl.gg/#/roles?id=reaction-roles) used for onboarding:
  * e.g. `/gentei-if set only:@readtherules`
* [carl-bot muterole](https://docs.carl.gg/#/moderation?id=punishment-commands): when carl-bot gives a user the mute role, Gentei can be made to revoke the role at the same time. 
  * e.g. `/gentei-if set only:@readtherules not:@muterole`
* Cooperating with non-Gentei membership roles: if users want to switch between membership bots, this helps keep the role grant list tidy.
  * e.g. `/gentei-if set not:@members (menshi),@members (janktei)`

## Testing

To allow you to test changes without trying things out in a test server or test user, `/gentei-if test` will have the same `only:` and `not:` options above and add the `has:` option. 

Populate the `has:` option with all of the roles that you want a test user to have - with the default being no roles - and Gentei will tell you whether a user with that role set would be eligible for a role. The plan is to have a :heavy_check_mark: and ❌ list to explain why.

## Relevant command changes

`/gentei info` and `/gentei check` will allow users to see the conditions for a server and how + if any of them apply. This should hopefully reduce/remove the amount of "why don't I have the role" questions by allowing admins and mods to tell users to run the command.

---

As always, I'm open to feedback! Feel free to opine or ask questions either in a comment below or the #gentei-限定 channel on Discord.

**Comment by mark-ignacio** [comment]
Design proposals/clarifications:
* For multi-role servers, these conditions apply to all Gentei-managed roles in the server.
* The `only:` and `not:` arguments will be evaluated like so:

```py
user_roles: set[int]
"Roles that the user in question currently holds"
only_roles: set[int]
"Roles specified in the `only:` argument"
not_roles: set[int]
"Roles specified in the `not:` argument"

if not only_roles and not not_roles:
    # neither set
    can_get_role = True
else:
    if only_roles:
        # overlapping set ("user has one of these roles")
        can_get_role = len(user_roles & only_roles) > 0
    if not_roles:
        # disjoint set ("user has exactly 0 of these roles")
        can_get_role &= user_roles.isdisjoint(not_roles)
```

The actual implementation will evaluate each role individually to ensure that a single code segment is responsible for evaluations for the entire program; this also allows the info + check commands to return rejection reasons.

**Comment by mark-ignacio** [comment]
`/gentei-if unset` would probably also be a good idea!

**Comment by brasstax** [comment]
This proposal seems good. What happens if a user re-runs the command, but with a different set of roles? Should it be additive, or overwrite the options entirely? I could see a scenario where a server implements menshi, jankei, etc after implementing gentei and then needing to rerun the command.

When setting the roles, do you plan on also having a role check after the role is run, or should it be up to the cronjob to run and then check against the list of only/not?

**Comment by mark-ignacio** [comment]
For clarity and simplicity's sake, re-running the set command will overwrite the existing configuration.

And yes, the cronjob will usually be the first thing that makes sure that _everyone_ in the server adheres to the new conditions. On an individual basis though, role change events for registered users will be the second trigger. As a particularly fun consequence of that for servers with a read-the-rules role, a Gentei-managed role will be granted shortly after registered + verified users complete server onboarding.
